### PR TITLE
Replace 'disabled' attributes from HTML anchor tags #7943

### LINKED ISSUE #7943: Replace 'disabled' attributes from HTML anchor tags
Noticed while working on #6634.

A lot of our HTML anchor tags (`<a>`) use the `disabled` attribute to disable clicks on the link. However, according to the dev docs, the `disabled` property is not compatible with the anchor tag. (Sources: [StackOverflow Question](https://stackoverflow.com/questions/18711317/is-disabled-a-valid-attribute-for-an-anchor-tag) and [W3 docs](https://dev.w3.org/html5/html-author/#the-a-element))

So far, I have found the following tag files use the disabled property erroneously:

- feedbackSessionActions.tag
- feedbackSessionPublishButton.tag
- expandCollapseButton.tag
- remindButton.tag
- studentsTable.tag
- studentList.tag
- feedbackResponseCommentRow.tag
- rowActions.tag

### PR DESCRIPTION
**Pull request by junkiattan**
Fixes #7943 

Replaced 'disabled' attributes from HTML anchor tags for the following tags:
- feedbackSessionActions.tag
- feedbackSessionPublishButton.tag
- expandCollapseButton.tag
- remindButton.tag
- studentsTable.tag
- studentList.tag
- feedbackResponseCommentRow.tag
- rowActions.tag

Did this by ultilizing the disabled property of bootstrap and returning false when click on the button.

### CODE REVIEW

**Comment by tshradheya** [comment]
@junkiattan If you are wondering, why the tests are failing and want an easy way to fix them. Have a read through [GodMode](https://github.com/TEAMMATES/teammates/blob/master/docs/godmode.md)

**Comment by junkiattan** [comment]
@tshradheya Thanks for the tip man. I will give it a try and let you know if I encounter any difficulties. Is the error due to my editions though? I cant seem to understand from the results of the report from Travis-CI

**Comment by tshradheya** [comment]
Well, they aren't exactly errors, but yeah they are caused due to your changes. 

So basically you have changed the DOM structure(HTML) in a few places. But the exact same changes haven't been made the test files which compare the DOM structure and hence the assertion errors.

Godmode provides an easy way to fix this and you can follow the steps quite easily in the [document](https://github.com/TEAMMATES/teammates/blob/master/docs/godmode.md)

Travis log from line 2625 will give you a better understanding

**Comment by junkiattan** [comment]
@whipermr5 @tshradheya

I ran into some issues. After making changes to the code and running the tests, I keep failing 4 test cases.

However after setting breakpoints and then debugging to see where the problem was by tracing the code, I realized that the UI was indeed working as expected.

For example, under the InstructorFeedbackSessionsPageUiTest.java ln 587, the check asserts whether the caret beside the results option was disabled. Running the test returns a false, however the actual UI shows that the actual caret is disabled.

My guess is that the test checks for the "disabled" attribute, which is definitely the case for error4, but not sure for 2,3. However since the disabled attributes are removed and an alternative was introduced to disable the button within the class, this was not read by the old tests.

Below are the 4 errors. Error1 relates to the presence of errors, hence would be most likely solved once errors 2,3,4 are solved

error1
error1

error2
error2

error3
error3

error4
error4

I would need some advice on how to proceed on from here. Do I change the tests also? Or would that be done in a separate issue?

**Comment by LiHaoTan** [comment]
Just to add my opinion, anchor tags are not semantically the same as buttons. So perhaps anchor tags with input/btn tags when they are actually buttons is better and also improves accessibility.

**Comment by whipermr5** [comment]
> I would need some advice on how to proceed on from here. Do I change the tests also? Or would that be done in a separate issue?

Yes; change the tests to check for the new disabled class.

File: src/main/webapp/WEB-INF/tags/instructor/feedbacks/feedbackSessionActions.tag [hunk 1]
```diff
@@ -3,22 +3,20 @@
 <%@ taglib tagdir="/WEB-INF/tags/instructor/feedbacks" prefix="tif" %>
 <%@ tag import="teammates.common.util.Const" %>
 <%@ attribute name="actions" type="teammates.ui.template.InstructorFeedbackSessionActions" required="true" %>
-<a class="btn btn-default btn-xs btn-tm-actions session-edit-for-test margin-bottom-7px"
+<a class="btn btn-default btn-xs btn-tm-actions session-edit-for-test margin-bottom-7px<c:if test="${not actions.allowedToEdit}"> disabled</c:if>"
     href="${actions.editLink}"
     title="<%= Const.Tooltips.FEEDBACK_SESSION_EDIT %>"
     data-toggle="tooltip"
-    data-placement="top"
-    <c:if test="${not actions.allowedToEdit}">disabled</c:if>>
```
Comment (wenmogu) [inline src/main/webapp/WEB-INF/tags/instructor/feedbacks/feedbackSessionActions.tag]: Hi friends just want to clarify (as a noob to jsp): so based on my understanding, the thing enclosed inside `${}` should be some object passed into this tag, and thus the `allowedToEdit` is one of the attributes of this `actions` enclosed inside `${}`. However based on my tracing, this action eventually goes back to `InstructorFeedbackSessionActions`, which has `editLink`, `remindLink`... and `isAllowedToEdit` as contrary to the `allowedToEdit` as accessed in the `${}` here. So can someone help me clarify where this `allowedToEdit` comes from? or is my understanding of jsp wrong? thx!

Comment (LiHaoTan) [reply src/main/webapp/WEB-INF/tags/instructor/feedbacks/feedbackSessionActions.tag]: That's a JavaBean, so `allowedToEdit` is matched by the pattern `isAllowedToEdit`

File: src/test/java/teammates/test/pageobjects/AppPage.java [hunk 2]
```diff
@@ -1,8 +1,9 @@
 package teammates.test.pageobjects;
 
+import static org.junit.Assert.assertNotNull;
+import static org.testng.Assert.assertTrue;
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: Let's revert all the changes of this file?

File: src/test/java/teammates/test/cases/browsertests/StudentHomePageUiTest.java [hunk 3]
```diff
@@ -158,8 +157,7 @@ private void testLinks() {
 
         ______TS("link: link of pending feedback");
 
-        assertEquals("true", studentHomePage.getViewFeedbackButton("First Feedback Session").getAttribute("disabled"));
-
+        assertEquals("true", studentHomePage.getViewFeedbackButton("First Feedback Session").getAttribute("disabled"));
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/browsertests/StudentHomePageUiTest.java]: Let's revert all the changes of this file.

File: src/test/resources/pages/godmode.html [hunk 4]
```diff
@@ -56,4 +56,4 @@ <h2>Unpredictable values</h2>
             </ul>
         </div>
     </body>
-</html>
+</html>
```
Comment (wkurniawan07) [inline src/test/resources/pages/godmode.html]: Unrelated change?

File: src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java [hunk 5]
```diff
@@ -225,9 +225,9 @@ private void testRemindActions() {
         ______TS("remind action: AWAITING feedback session");
 
         homePage.verifyUnclickable(homePage.getRemindLink(feedbackSessionAwaiting.getCourseId(),
-                                                          feedbackSessionAwaiting.getFeedbackSessionName()));
+                feedbackSessionAwaiting.getFeedbackSessionName()));
         homePage.verifyUnclickable(homePage.getRemindOptionsLink(feedbackSessionAwaiting.getCourseId(),
-                                                                 feedbackSessionAwaiting.getFeedbackSessionName()));
+                feedbackSessionAwaiting.getFeedbackSessionName()));
```
[the remark this answers is not in the document]
Comment (junkiattan) [reply src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: @wkurniawan07 

What you mean is to revert undo everything I have done for all files and just redo using the button tags?

[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: For most files, yes. You should not need to revert the test HTML files because if you do you'd have to change them again with God mode anyway.

**Comment by junkiattan** [comment]
@wkurniawan07 I have done as suggested. but there are some assertion errors in my local build.

Apparently changing to button tags affects how some of the tests function as well. Some of the assertions are looking for a certain string on the webpage but it turns out the string is either not there or that the code now reads from the wrong position. 

I do not understand how changing from a link to a button can lead to such errors. Do you have any advice?

@whipermr5 @LiHaoTan

**Comment by wkurniawan07** [comment]
@junkiattan I checked the Travis build log and can identify some problems:
- It looks like you're running God mode with outdated JS file. You should merge/rebase with the latest `master` branch, run `npm run build`, and use that build to run God mode.
- Some tests rely on the fact that the element is an `a` rather than a `button`, in which case you do need to change those tests accordingly.

**Comment by junkiattan** [comment]
@wkurniawan07 

My branch is up to date with master when I ran Godmode. I re-ran the npm run build and rebuild using godmode alr.

I have tried to identify the tests, but some of those areas, such as those involving fillRichTextEditor in AppPage.java and getNewPageInstance in AppPage.java.

Would appreciate some assistance in this area. I am not too familiar with the how the tests were constructed. I have already identified some of the tests which checked for <a> tags and replaced them by tracing the code.

**Comment by wkurniawan07** [comment]
I just realized a grave consequence of converting everything to button: `button` tags do not recognize `href` attribute. That means all the expected "click here to go to another page" will fail, and that is the source of almost all the failures you see right now.
We need to implement either:
```html
<button>
  <a href="...">Text</a>
</button>
```
Or keep it at status quo (using `a`) and changing `disabled` attribute to `disabled` class, just like how the issue is meant to be dealt with from the beginning. I actually prefer the latter (i.e. keep at `a`). @LiHaoTan may I know what are the caveats that make you think this is not a good practice? (I know it's not but I want to know your reasoning such that you want to make this radical change)

**Comment by LiHaoTan** [comment]
@wkurniawan07

I think the solution we should be either:
* Keep the “direct link buttons” as anchor tags and disable them properly  (https://css-tricks.com/how-to-disable-links/) OR
* Make all the "direct link buttons” and add a click handler to them.

I prefer the latter approach because we can actually properly disable buttons unless it takes substantially more effort to implement.

In that case, then I believe we need to at least to do the former of disabling anchors as much as we can. The former probably can reach parity with buttons as the web progresses.

**Rationale**

The anchor tags have the downside of not being able to be properly disabled.

Although it has improved a lot over the years there are still some issues like support for `pointer-events` which is not yet available in all major browsers.

Then another headache is that they still cause some problems with accessibility such as the different ways of activating anchors vs buttons but I guess we could take the idea that many people already know the gotchas and can work with them.

Which makes seems like buttons are superior, but they don't have `href` so we need to write our own click handler. But then again if they have `href` how can they be disabled? If they can be disabled then why not allow disabling anchors?

So without anchors, buttons do not support progressive enhancement. But at least we can say with the design TEAMMATES currently have we don't have the resources to support that.

So I think whether we agree with how the specs ended up with such a design is not something we should attempt to overturn because it comes with a lot of gotchas. I know they have discussed this issue to death and button is still not going to get `href` and anchors are still not going to get `disabled`.

Because the specs did not want to allow disabling anchors it gets [extremely complicated to disable anchors](https://css-tricks.com/how-to-disable-links/). (I know the article is one sided but it tells us how to disable anchors (almost) completely)

So in this case we have direct links, and they feel more like links than buttons, but we are going to have some trouble disabling them.

Or alternatively, we can change the meaning of our “direct link buttons” semantically and make them just buttons that perform actions and not links, whether they navigate to a link or send an AJAX request is irrelevant to the user. This requires changing the “direct link buttons” to use a `click` handler.

**Comment by junkiattan** [comment]
I think that the best course of action would be to move this discussion to a new issue.

No matter the advantage buttons has over links, it doesnt make sense to change some links (and their respective tests) to buttons now and leave the rest untouched as it will affect future prs that will touched related methods and functionality. On the other hand, changing all to buttons (if this is the chosen action) is out of the scope of this issue and should definitely be discussed further as it touches on the foundations of the project. It is apparant from the constant chnage in implementation direction requested of me that we should move this to a separate issue to hold further discussions on what should be done about the tags.

I have discussed this matter with @whipermr5 and we have agreed on this.

**Comment by LiHaoTan** [comment]
@junkiattan Let's go with whatever that's the simplest to execute right now. Sorry for all the work you have to go through. We can address the real problem later on in a separate issue.

File: src/test/java/teammates/test/cases/browsertests/StudentHomePageUiTest.java [hunk 6]
```diff
@@ -102,6 +102,7 @@ private void testContentAndLogin() throws Exception {
         studentHomePage.verifyHtmlMainContent("/studentHomeTypicalTestingSanitization.html");
     }
 
+    @SuppressWarnings("PMD.UseAssertTrueInsteadOfAssertEquals")
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/browsertests/StudentHomePageUiTest.java]: If this is necessary, can add a comment explaining why?

Comment (junkiattan) [reply src/test/java/teammates/test/cases/browsertests/StudentHomePageUiTest.java]: Hey john, I realized that this ccould be replaced with assertTrue instead, hence removing the need for this suppression. Will push again.

File: src/test/java/teammates/test/pageobjects/AppPage.java [hunk 7]
```diff
@@ -795,7 +796,11 @@ public boolean isElementCovered(WebElement element) {
     }
 
     public void verifyUnclickable(WebElement element) {
-        assertNotNull(element.getAttribute("disabled"));
+        try {
+            assertTrue(element.getAttribute("class").contains("disabled"));
+        } catch (AssertionError e) {
+            assertNotNull(element.getAttribute("disabled"));
+        }
```
Comment (whipermr5) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: This logic would be better: if element is an anchor, check for class, else check for attribute.

File: src/test/java/teammates/test/pageobjects/AppPage.java [hunk 8]
```diff
@@ -796,9 +796,10 @@ public boolean isElementCovered(WebElement element) {
     }
 
     public void verifyUnclickable(WebElement element) {
-        try {
+        if (element.getTagName().contains("a")){
```
Comment (whipermr5) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: `.equals`?

File: src/main/webapp/WEB-INF/tags/instructor/feedbacks/feedbackSessionActions.tag [hunk 9]
```diff
@@ -56,10 +53,9 @@
       Results
     </a>
     <button type="button"
-        class="btn btn-default btn-xs btn-tm-actions dropdown-toggle session-results-options"
+        class="btn btn-default btn-xs btn-tm-actions dropdown-toggle session-results-options<c:if test="${not actions.publishButton.actionAllowed || actions.privateSession}"> disabled</c:if>"
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/feedbacks/feedbackSessionActions.tag]: We're not doing this for buttons right?

File: src/main/webapp/WEB-INF/tags/instructor/feedbacks/feedbackSessionActions.tag [hunk 10]
```diff
@@ -77,17 +73,15 @@
...
+    <a class="btn btn-default btn-xs btn-tm-actions session-remind-for-test<c:if test="${not actions.allowedToRemind || actions.privateSession}"> disabled</c:if>"
         href="${actions.remindLink}"
-        data-fsname="${actions.fsName}"
-        <c:if test="${not actions.allowedToRemind || actions.privateSession}">disabled</c:if>>
+        data-fsname="${actions.fsName}">
       Remind
     </a>
     <button type="button"
-        class="btn btn-default btn-xs btn-tm-actions dropdown-toggle session-remind-options-for-test"
+        class="btn btn-default btn-xs btn-tm-actions dropdown-toggle session-remind-options-for-test<c:if test="${not actions.allowedToRemind || actions.privateSession}"> disabled</c:if>"
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/feedbacks/feedbackSessionActions.tag]: Same here.

**Comment by junkiattan** [comment]
@wkurniawan07 

What do you mean by 

> Simply remove the class and click the button

@whipermr5

**Comment by whipermr5** [comment]
I think he means through using the browser development tools that enable DOM manipulation? The tools also allow the user to remove the `disabled` attribute though.

File: src/main/webapp/WEB-INF/tags/student/home/rowActions.tag [hunk 11]
```diff
@@ -3,15 +3,14 @@
 <%@ tag import="teammates.common.util.Const" %>
 <%@ attribute name="actions" type="teammates.ui.template.StudentFeedbackSessionActions" required="true" %>
 <%@ attribute name="index" required="true" %>
-<a class="btn btn-default btn-xs btn-tm-actions"
+<a class="btn btn-default btn-xs btn-tm-actions<c:if test="${not actions.sessionPublished}"> disabled</c:if>"
     href="${actions.sessionPublished ? actions.studentFeedbackResultsLink : 'javascript:;'}"
```
Comment (wkurniawan07) [inline src/main/webapp/WEB-INF/tags/student/home/rowActions.tag]: @junkiattan like this

File: src/test/resources/pages/studentHomeHTML.html [hunk 12]
```diff
@@ -31,6 +31,29 @@ <h1>
   </div>
   <br>
   <br>
+  <div class="panel panel-primary">
```
[the remark this answers is not in the document]
Comment (junkiattan) [reply src/test/resources/pages/studentHomeHTML.html]: Hm that's very strange. Perhaps it was inserted from another merged PR when I updated the branch earlier, because I definitely did not insert this part through my changes

File: src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentRow.tag [hunk 13]
```diff
@@ -43,7 +43,7 @@
     <div class="col-xs-2">
       <c:if test="${frc.editDeleteEnabled}">
         <form class="responseCommentDeleteForm pull-right">
-          <a href="<%= Const.ActionURIs.INSTRUCTOR_FEEDBACK_RESPONSE_COMMENT_DELETE %>"
+          <a href="${frc.editDeleteEnabled ? <%= Const.ActionURIs.INSTRUCTOR_FEEDBACK_RESPONSE_COMMENT_DELETE %> : 'javascript:;'}"
```
[the remark this answers is not in the document]
Comment (junkiattan) [reply src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentRow.tag]: thanx for the valuable tip! I was wondering whether it was right to right my original statement and was suspecting the errors came from there but did not know how to change it.

File: src/main/webapp/WEB-INF/tags/instructor/results/remindButton.tag [hunk 14]
```diff
@@ -6,8 +6,8 @@
 <%@ attribute name="remindButton" type="teammates.ui.template.InstructorFeedbackResultsRemindButton" required="true" %>
 
 <div class="remind-no-response">
-  <a href="javascript:;" data-actionlink="${remindButton.urlLink}" class="${remindButton.className}" data-toggle="modal"
-      <c:if test="${remindButton.disabled}">disabled=""</c:if> data-target="#remindModal">
+  <a href="javascript:;" data-actionlink="${remindButton.urlLink}" class="${remindButton.className}<c:if test="${remindButton.disabled}"> disabled</c:if>" data-toggle="modal"
```
Comment (wkurniawan07) [inline src/main/webapp/WEB-INF/tags/instructor/results/remindButton.tag]: This one is quite a unique case where the link is in the `data-actionlink` instead of `href`, but let's try the same treatment.

File: src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentRow.tag [hunk 15]
```diff
@@ -43,16 +43,14 @@
     <div class="col-xs-2">
       <c:if test="${frc.editDeleteEnabled}">
         <form class="responseCommentDeleteForm pull-right">
-          <a href="<%= Const.ActionURIs.INSTRUCTOR_FEEDBACK_RESPONSE_COMMENT_DELETE %>"
+          <a <c:set var="deleteUri" value="<%= Const.ActionURIs.INSTRUCTOR_FEEDBACK_RESPONSE_COMMENT_DELETE %>" />
```
Comment (wkurniawan07) [inline src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentRow.tag]: While it works, it doesn't really look clean to set a JSP variable inside an HTML tag. Let's move it outside.

**Comment by junkiattan** [comment]
@wkurniawan07 @whipermr5 

For review please;)

File: src/test/resources/pages/godmode.html (hunk 4 above)
Comment (wkurniawan07) [reply src/test/resources/pages/godmode.html]: Left with this...

Comment (junkiattan) [reply src/test/resources/pages/godmode.html]: Hey @wkurniawan07 ,

I am puzzled about how to solve this. Is it just an issue of adding "\n"? Plus I didn't change this file at all for my issue, don't know why there is a change here in the first place.

Comment (wkurniawan07) [reply src/test/resources/pages/godmode.html]: > Is it just an issue of adding `\n`?

It is. It's not visible in the "Conversation" view but in the "Files changed" view the diff is displayed like that.

File: src/test/resources/pages/godmode.html [hunk 16]
```diff
@@ -56,4 +56,4 @@ <h2>Unpredictable values</h2>
             </ul>
         </div>
     </body>
-</html>
+</html>\n
```
Comment (junkiattan) [inline src/test/resources/pages/godmode.html]: @wkurniawan07 

It seems that "\n" is not working. I am trying it with "<br>" instead

File: src/test/resources/pages/godmode.html [hunk 17]
```diff
@@ -56,4 +56,4 @@ <h2>Unpredictable values</h2>
             </ul>
         </div>
     </body>
-</html>\n
\ No newline at end of file
+</html><br>
```
Comment (whipermr5) [inline src/test/resources/pages/godmode.html]: Use a newline character instead

File: src/test/resources/pages/godmode.html (hunk 16 above)
Comment (whipermr5) [inline src/test/resources/pages/godmode.html]: The newline character is what is inserted when you press enter/return.

**Comment by damithc** [comment]
Looks like this PR required lot of work over a long period. Kudos to everyone who contributed in different ways. 👍

**Comment by whipermr5** [comment]
@junkiattan Oops, there's a new conflict in the remind button tag after I merged #8581. Can add the tooltip to your code and mark it as resolved? Will merge after that.

**Comment by junkiattan** [comment]
@whipermr5 

You can merge now;)
